src/algorithms/simulation/simple_simulation.cpp

```diff
@@ -0,0 +1,81 @@
+#include "algorithms/simulation/simple_simulation.hpp"
+
+#include "core/gate.hpp"
+#include "core/target_tags.hpp"
+#include "core/utils/timer.hpp"
+
+namespace syrec {
+
+    boost::dynamic_bitset<>& core_gate_simulation::operator()(const gate& g, boost::dynamic_bitset<>& input) const {
+        if (is_toffoli(g)) {
+            boost::dynamic_bitset<> c_mask(input.size());
+            for (gate::const_iterator itControl = g.begin_controls(); itControl != g.end_controls(); ++itControl) {
+                c_mask.set(*itControl);
+            }
+
+            if (c_mask.none() || ((input & c_mask) == c_mask)) {
+                input.flip(*g.begin_targets());
+            }
+
+            return input;
+        } else if (is_fredkin(g)) {
+            boost::dynamic_bitset<> c_mask(input.size());
+            for (gate::const_iterator itControl = g.begin_controls(); itControl != g.end_controls(); ++itControl) {
+                c_mask.set(*itControl);
+            }
+
+            if (c_mask.none() || ((input & c_mask) == c_mask)) {
+                // get both positions and values
+                gate::const_iterator it = g.begin_targets();
+                unsigned             t1 = *it++;
+                unsigned             t2 = *it;
+
+                bool t1v = input.test(t1);
+                bool t2v = input.test(t2);
+
+                // only swap when different
+                if (t1v != t2v) {
+                    input.set(t1, t2v);
+                    input.set(t2, t1v);
+                }
+            }
+
+            return input;
+        } else {
+            assert(false);
+        }
+        return input;
+    }
+
+    bool simple_simulation(boost::dynamic_bitset<>& output, circuit::const_iterator first, circuit::const_iterator last, const boost::dynamic_bitset<>& input,
+                           const properties::ptr& settings,
+                           const properties::ptr& statistics) {
+        auto gate_simulation = get<gate_simulation_func>(settings, "gate_simulation", core_gate_simulation());
+
+        timer<properties_timer> t;
```

Comment:
I guess the reason why it is constructed outside statistic block is because the simulation should happen irrespective of the statistics. As a result, it is constructed outside and its can be used to used stop when the simulation gets over.

Reply:
Oh yeah. Forget the comment above.